src/ray/object_manager/object_manager.cc

```diff
@@ -1,156 +1,311 @@
+  }
+  return status;
+};
+
+ray::Status ObjectManager::TransferCompleted() {
+  // RAY_LOG(INFO) << "TransferCompleted " << client_id_;
+  num_transfers_ -= 1;
+  return DequeueTransfers();
+};
+
+ray::Status ObjectManager::ExecuteSend(const ObjectID &object_id,
+                                       const ClientID &client_id) {
+  //  RAY_LOG(INFO) << "ExecuteSend "
+  //                << object_id << " "
+  //                << client_id << " ";
   ray::Status status;
-  status =
-      GetTransferConnection(client_id, [this, object_id](SenderConnection::pointer conn) {
-        ray::Status status = QueuePush(object_id, conn);
-      });
+  status = connection_pool_.GetSender(ConnectionPool::TRANSFER, client_id,
+                                      [this, object_id](SenderConnection::pointer conn) {
+                                        // TODO(hme): Handle this error. It's
+                                        // possible that the object is not local,
+                                        // so this should not be a fatal error.
+                                        RAY_CHECK_OK(SendHeaders(object_id, conn));
+                                      },
+                                      [this, object_id]() {
+                                        // Push is best effort, so do nothing on failure.
+                                      });
   return status;
+}
+
+ray::Status ObjectManager::SendHeaders(const ObjectID &object_id_const,
+                                       SenderConnection::pointer conn) {
+  // RAY_LOG(INFO) << "SendHeaders";
+  ObjectID object_id = ObjectID(object_id_const);
+  // Allocate and append the request to the transfer queue.
+  plasma::ObjectBuffer object_buffer;
+  plasma::ObjectID plasma_id = object_id.to_plasma_id();
+  std::shared_ptr<plasma::PlasmaClient> store_client = store_pool_->GetObjectStore();
+  ARROW_CHECK_OK(store_client->Get(&plasma_id, 1, 0, &object_buffer));
+  if (object_buffer.data_size == -1) {
+    RAY_LOG(ERROR) << "Failed to get object";
+    // If the object wasn't locally available, exit immediately. If the object
+    // later appears locally, the requesting plasma manager should request the
+    // transfer again.
+    (void)connection_pool_.ReleaseSender(ConnectionPool::TRANSFER, conn);
+    return ray::Status::IOError(
+        "Unable to transfer object to requesting plasma manager, object not local.");
+  }
+  RAY_CHECK(object_buffer.metadata->data() ==
+            object_buffer.data->data() + object_buffer.data_size);
+
+  TransferQueue::SendContext context;
+  context.client_id = conn->GetClientID();
+  context.object_id = object_id;
+  context.object_size = object_buffer.data_size;
+  context.data = const_cast<uint8_t *>(object_buffer.data->data());
+  UniqueID context_id = transfer_queue_.AddContext(context);
+
+  // Create buffer.
+  flatbuffers::FlatBufferBuilder fbb;
+  auto message = CreatePushRequestMessage(fbb, fbb.CreateString(object_id.binary()),
```

Comment:
Isn't this what the `to_flatbuf` function is for?

https://github.com/ray-project/ray/blob/f69cbd35d4e86f2a3c2ace875aaf8166edb69f5d/src/common/common_protocol.h#L12-L19

Reply:
Wasn't aware of this; though currently I'm getting compile-time errors when I include `common_protocol.h`. Added a TODO for this, which will entail resolving the import issue.